##'  Analysis of indels 7.12.2023 
#'  
#' scp filip_kolar@storage-brno12-cerit.metacentrum.cz:brno3/ERC/arenosa_mapped_to_arenosa/filtering_results/indels/pop.5x.ac0.indels.annotated/genotype.counts.norm.tsv .
#'
#' The data were generated by scripts /storage/brno12-cerit/home/filip_kolar/brno3/scripts/load
#' 
#'  filter_indels.sh >  pop.snpeff.sh > count_hom_het_indels.sh
#'  In the first step reliable indels passing GATK best pract. were selected and 
#'  Only indels supported by 5x and more are considered.
#'  This differs from the september analysis because I found that sfs from 8x
#'  dataset have deficiency of singletons 
#'  
#'  Then I split vcf per population, select indels with ABS(ILEN) > 20
#'  and subset only the 6 best individuals per population.
#'  
#'  Then classify indels into fitness categories by SnpEff, 
#'  I will probably work only with the two categoris HIGH effect which are mainly frameshift mutations in CDS
#'  
#'  The neutral category MODIFIER hits mainly regions away from CDS. 
#'    
#' remove populations with less than 6 inds: 
#' 
#' grep -v "less" genotype.counts.norm.tsv > genotype.counts.tsv
getwd()
setwd("/home/vlkofly/Arabidopsis_science/arenosa/indels/analysis_dec2023_norm")
library(tidyverse)
library(lme4)
library(ggrepel)
library(scales)

th<-theme(axis.text.x=element_text(size=23),
          axis.title = element_text(size=23),
          axis.text.y=element_text(size=23),plot.title=element_text(size=25),
          legend.text = element_text(size=18))

indpop<-read.table("genotype.counts.tsv",header=T,sep="\t")
indpop$ploidy<-as.factor(as.character(indpop$ploidy))
summary(indpop)
#indpop<-indpop[!is.na(indpop$dp),] # remove SWD 
indpop$pop<-as.factor(gsub(".scaffold_.*","",indpop$pop))# fix the population name
indpop$scf<-as.factor(as.character(gsub("scaffold_","",indpop$scf)))
#' the load values were calculated per chromosome, sum it here, only later maybe plot separate chromosomes
#' 

#' check the variance between the scaffolds
#' 
#' Plot the number of sites per scaffold:
svg("sites.per.scaffold.svg")
ggplot(indpop, aes(y=sites_tot/1000,x=scf,colour=ploidy))+
  geom_boxplot()+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  
  ylab("Number of indels/1e3")+
  th
dev.off()
#' **There are more sites in tetraploids**
#' 
#' This has changed between 5x and 8x dataset

#' Plot depth per scaffold
svg("depth.per.scaffold.svg")
ggplot(indpop, aes(y=dp,x=scf,colour=ploidy))+
  geom_boxplot()+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  
  th
dev.off()
#' There are more sites in tets although they in general show lower depth
#' 

#' Check the effect of depth on the number of indels per scaffold
#' 
svg("indel.No.depth.ploidyperscf.svg")
ggplot(indpop ,aes(x=sites_tot/10000,y=dp,colour=ploidy))+
  geom_point(size=4,pch=16)+
  geom_text(aes(label=pop),size=2.5,check_overlap = F,nudge_y = -2)+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  xlab("Number of indels/1e4")+
  th
dev.off()
#' Notice the tetraploid tail at the left bottom corner
#' 
#' This became less prominent in the 5x dataset
#' 


names(indpop)
# Let's sum the scaffold values to have a whole genome indices

gen.indpop<- indpop %>% group_by(pop) %>%
  summarise_if(is.numeric,sum) # this would sum all numeric variables, non-numeric are lost

# also for mean depth and number of individuals sum does not make sense, you have to go back to the average so divide by 8
gen.indpop$nind<-gen.indpop$nind/8
gen.indpop$dp<-gen.indpop$dp/8
gen.indpop$dp_raw<-gen.indpop$dp_raw/8

# and add ploidy
gen.indpop<-merge(gen.indpop,indpop[indpop$scf == "1",c("pop","ploidy")],by="pop")
summary(gen.indpop)

#' I included the actual dp of the dataset where the best individuals were subsetted
ggplot(gen.indpop,aes(x=dp_raw,y=dp,colour=ploidy))+
  geom_boxplot()+
  th

#' Check the effect of depth and remove low depth
#' The black line marks cutoff threshold
svg("indel.No.depth.ploidy.y20.y50.all.svg")

ggplot(gen.indpop,aes(x=sites_tot/10000,y=dp,colour=ploidy))+
  geom_point(size=3,pch=16)+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text(aes(label=pop),size=2.5,check_overlap = F,nudge_y = -2)+
  geom_hline(yintercept=20,colour="red")+
  geom_hline(yintercept=50,colour="red")+
  #geom_hline(yintercept=22,colour="blue")+
  xlab("Number of indels/1e4")+
  ylab("Depth of coverage")+
  th
dev.off()
#' as I used lower threshold to include the indel in, then I should probably use different 
#' population threshold, I should decrease it? Check what populations you used for the diversity?
#' 
#' Maybe I do not need as strict threshold for the population depth if I use it as covariate
#' Check the sfs to decide what populations to include
#' Also the dp here in the datset is the dp of the whole vcf before the 6 best individuals
#' have been selected redo it with actual depth
#' 
#' Well after checking the sfs I think I can select quite low coverage diploids
#' but tets must be above 25 because otherwise the sfs is disrupted by filtering
#' So I think I will select 11 for diploids and 22 for tetraploids
#' OK the normalized data diploids have lower depth I just have to select the same pops as in the previous approach


svg("indel.No.depth.ploidy.dp15.svg")
ggplot(gen.indpop[gen.indpop$dp>15,],aes(x=sites_tot/10000,y=dp,colour=ploidy))+
  geom_point(size=4,pch=16)+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  
  geom_text(aes(label=pop),size=2.5,check_overlap = F,nudge_y = -2)+
  xlab("Number of indels/1e4")+
  ylab("Depth of coverage")+
  th
dev.off()


#' So for subsequent analysis I choose only populations with dp between 20 and 50.
#' So I think I will select 11 for diploids and 22 for tetraploids
#' No I will just select 20
# gen.indpop20<- gen.indpop %>% 
#  filter(ploidy == 2 & dp > 15 & dp < 50 | ploidy == 4 & dp > 22 )
# summary(gen.indpop20)
pop20<-read.table("pop20.txt",col.names = "pop")
gen.indpop20<- gen.indpop %>% 
   filter(pop %in% pop20$pop)

summary(gen.indpop20)   

svg("indel.No.depth.ploidy.subset.svg")
ggplot(gen.indpop20,aes(x=sites_tot/10000,y=dp,colour=ploidy))+
  geom_point(size=4,pch=16)+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  
  geom_text(aes(label=pop),size=2.5,check_overlap = F,nudge_y = -2)+
  xlab("Number of indels/1e4")+
  ylab("Depth of coverage")+
  th
dev.off()

#' Should I also remove SUB? as it is an outlier and it breaks the pattern
#' 
#' Moreover it is basically the same population as BAB so I'm removing it
#' 
#' Like this I have 12 diploid and 16 tetraploid populations to work with

svg("depth.ploidy.subset.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=dp,fill=ploidy))+
  geom_boxplot()+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  ylab("Depth of coverage")+
  geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  th
dev.off()

dp.wilcox<-wilcox.test(gen.indpop20[gen.indpop20$ploidy==2,"dp"],gen.indpop20[gen.indpop20$ploidy==4,"dp"])
print(dp.wilcox) 

#' **Is there a difference in total number of indels between diploids and tetraploids?**

svg("total.indel.ploidy.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=sites_tot/10000,fill=ploidy))+
  ylab("Number of indels/1e4")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_boxplot()+
  geom_text(aes(label=pop),size=4,check_overlap = T,nudge_x = 0.2)+
    geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  th
dev.off()
#' First I will test the difference with Wilcox test
tot.wilcox<-wilcox.test(gen.indpop20[gen.indpop20$ploidy==2,"sites_tot"],gen.indpop20[gen.indpop20$ploidy==4,"sites_tot"])
print(tot.wilcox) 
#' The difference between the cytotypes is statistically significant
#' 
#' But these numbers of indels are actually all numbers considering also major alleles
#' 
#' It might be a problem to count also major alleles because most of them might be ancestral
#' 
#' And I want to count only derived alleles
#' 
#' For that purpose I selected only minor alleles --max-ac 6/12 diploid/tetraploid

svg("total.minor.indel.ploidy.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=minor_sites_tot/10000,fill=ploidy))+
  ylab("Number of indels/1e4")+
  geom_boxplot()+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text(aes(label=pop),size=4,check_overlap = T,nudge_x = 0.2)+
  geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  th
dev.off()
#' The difference gets even more pronounced
#' 
#' What is actually the distribution of the number of indels?
svg("distribution.minor.indels.svg")
ggplot(gen.indpop20,aes(x=minor_sites_tot/1e4,fill=ploidy))+
  xlab("Number of indels/1e4")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  
  geom_density(alpha=.5)+
  th
dev.off()
#' This does not seem like gausiann distr. lets use poisson in the models
#' I would like to test the effect of ploidy with depth as a covariate?
sites_tot.glm<-glm(minor_sites_tot~ploidy,data=gen.indpop20, family="poisson")
summary(sites_tot.glm) 
#' This is highly significant

sites_tot_dp.glm<-glm(minor_sites_tot~ploidy*dp,data=gen.indpop20, family="poisson")
summary(sites_tot_dp.glm) 
#' This is significant as well
#' 
#' Is the model with dp better?
anova(sites_tot.glm,sites_tot_dp.glm,test="LRT") # what test to use with poisson? 
#' Of course the model with explains much more, but still the effect of ploidy is significant
#' I should consider the effect of depth in each step of the analysis
#' 
#' Filip suggested to use dp as random factor in glmer

sites_tot_dp.glmer<-glmer(minor_sites_tot~ploidy+(1|dp),data=gen.indpop20, family="poisson")
summary(sites_tot_dp.glmer) 

#'
#'  
#' How does the effect of dp look in minor alleles?
#' 
svg("minor.indel.depth.ploidy.svg")
ggplot(gen.indpop20,aes(x=minor_sites_tot/1e4,y=dp,colour=ploidy))+
  geom_point(size=4,pch=16)+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  
  geom_text(aes(label=pop),size=2.5,check_overlap = F,nudge_y = -1)+
  xlab("Number of indels/1e4")+
  ylab("Depth of coverage")+
  th
dev.off()
#' The effect of depth seems actually less pronounced in minor indels
#' 
#' Conclusion from the  total indel analysis is that tetraploids accumulate on average more indels than diploids.
#'
#' Mean values per ploidy: 
sum.per.ploidy<-gen.indpop20 %>%
  group_by(ploidy)%>%
  summarise(across(where(is.numeric),list(mean=mean,sd=sd,median=median))) %>%
  as.matrix() %>% t() %>% as.data.frame()
names(sum.per.ploidy) <- c("diploid","tetraploid")
print(sum.per.ploidy)
#' There is not much difference in mean depth 32.04/34.20 and its standard deviation.
#' 
#' Mean total number of indels is 1.73M/2.42M
#' Mean total number of minor indels is 1.39M/2.14M
#'___________________________________________________________________
#'
##' Let's focus on the two fitness categories of indels MODIFIER and HIGH
#'  The point is to test whether tetraploids not only accumulate more of indels
#'  but also if they accumulate more deleterious indels which are in the HIGH
#'  category. For this analysis one must take the assumption that the total numbers
#'  of HIGH indels are only negligibly affected by adaptive selection.
#'  I think it is well satisfied assumption given that I work with only minor allele indels 
#'  
#'  Let's see the distribution of HIGH and MODIF indels per ploidy
head(gen.indpop20)
#' High effect first, they are of course in minority
svg("high.N.indel.ploidy.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=HIGH_tot_minor/1e4,fill=ploidy))+
  ylab("Number of high effect indels/1e4")+
  geom_boxplot()+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  geom_text(aes(label=pop),size=3,check_overlap = T,nudge_x = 0.2)+
  th
dev.off()
#' ZIT, PER, HRN and VYR quite overlap with the diploid populations
#' 
#' But overall the effect of ploidy is evident even in highly deleterious indels
#' 
#' Here I have an idea of counting the load per individual and get also variance at within population level
#' 
#' What if I check the dataset also with major alleles?
#' 
svg("high.N.all.indel.ploidy.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=HIGH_tot/1e4,fill=ploidy))+
  ylab("Number of high effect indels/1e4")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  geom_boxplot()+
  geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  geom_text(aes(label=pop),size=4,check_overlap = T,nudge_x = 0.2)+
  th
dev.off()
#' There is not much difference, with minor alleles I lost ~5000 sites
#' 
#' Modifier alleles: 
#' 
svg("modif.N.indel.ploidy.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=MODIF_tot_minor/1e4,fill=ploidy))+
  ylab("Number of weak effect indels/1e4")+
  geom_boxplot()+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  
  geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  geom_text(aes(label=pop),size=4,check_overlap = T,nudge_x = 0.2)+
  th
dev.off()
#' There is wider spread in diploids, but still the difference is clear.
#' 
#' Is it statistically significant?
modif.wilcox<-wilcox.test(gen.indpop20[gen.indpop20$ploidy==2,"MODIF_tot_minor"],gen.indpop20[gen.indpop20$ploidy==4,"MODIF_tot"])
print(modif.wilcox)
high.wilcox<-wilcox.test(gen.indpop20[gen.indpop20$ploidy==2,"HIGH_tot_minor"],gen.indpop20[gen.indpop20$ploidy==4,"HIGH_tot"])
print(high.wilcox)
#' Yes of course, now include dp as covariate
sites_high_dp.glm<-glm(HIGH_tot_minor~ploidy*dp,data=gen.indpop20, family="poisson")
summary(sites_high_dp.glm)
#' It is still significant, but only with p=0.0205 
sites_low_dp.glm<-glm(MODIF_tot_minor~ploidy*dp,data=gen.indpop20, family="poisson")
summary(sites_low_dp.glm)
#' This is highly significant
#' Now I want to understand if the tets accumulate proportionally more higly deleterious
#' First check the ratio high.minor/total.minor?
gen.indpop20$hm.tm<-gen.indpop20$HIGH_tot_minor/gen.indpop20$minor_sites_tot
svg("high.tot.minor.ratio.ploidy.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=hm.tm,fill=ploidy))+
  ylab("High/total indels")+
  geom_boxplot()+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  geom_text(aes(label=pop),size=4,check_overlap = F,nudge_x = 0.2)+
  th
dev.off()
#' This ration seems to be higher in diploids, do they accumulate proportionally more high effect indels?
#' Is this difference significant?
svg("distribution.hm.tm.ratio.svg")
ggplot(gen.indpop20,aes(x=hm.tm,fill=ploidy))+
  xlab("High/Total indels")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  
  geom_density(alpha=.5)+
  th
dev.off()

ratio.hm.tm_dp.glm<-glm(hm.tm~ploidy*dp,data=gen.indpop20, family="gaussian")
summary(ratio.hm.tm_dp.glm)

#' This is not significant in glm 

hm.tm.wilcox<-wilcox.test(gen.indpop20[gen.indpop20$ploidy==2,"hm.tm"],gen.indpop20[gen.indpop20$ploidy==4,"hm.tm"])
print(hm.tm.wilcox)

#' And Wilcox is not significant either

#' I guess it will be the same for high/weak effect indels?
#' 
gen.indpop20$hm.wm<-gen.indpop20$HIGH_tot_minor/gen.indpop20$MODIF_tot_minor

svg("high.weak.ratio.ploidy.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=hm.wm,fill=ploidy))+
  ylab("High/Weak Indels")+
  geom_boxplot()+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  
  geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  geom_text(aes(label=pop),size=4,check_overlap = T,nudge_x = 0.2)+
  th
dev.off()

ratio.hm.wm_dp.glm<-glm(hm.wm~ploidy*dp,data=gen.indpop20, family="gaussian")
summary(ratio.hm.wm_dp.glm)
#' This is not significant
#' 
#' Well this ratio is hard to compare between ploidies as written somewhere in Conover 2022 article
#' 
#' Maybe it would help to focus only on CDS and get some neutral category there ... introns
#' 

gen.indpop20$hm.im<-gen.indpop20$HIGH_tot_minor/gen.indpop20$INTRON_tot_minor

 
svg("high.intron.ratio.ploidy.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=hm.im,fill=ploidy))+
  ylab("High/Weak Intron Indels")+
  geom_boxplot()+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  
  geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  geom_text(aes(label=pop),size=4,check_overlap = T,nudge_x = 0.2)+
  th
dev.off()

ratio.hm.im_dp.glm<-glm(hm.im~ploidy*dp,data=gen.indpop20, family="gaussian")
summary(ratio.hm.im_dp.glm)
#' It is slightly more significant, but not below 0.05
#' Wilcox?
hm.im.wilcox<-wilcox.test(gen.indpop20[gen.indpop20$ploidy==2,"hm.im"],gen.indpop20[gen.indpop20$ploidy==4,"hm.im"])
print(hm.im.wilcox)
#' Wilcox is not significant
#' 
#' **I will try a different approach**
#' I take the number of weak indels as a background rate of indel insertion
#' 
#' Compare if an increase in weak indels leads to proportional increase in high effect indels
#' 
#' If the slope of correlation between high and weak indels is steeper in tets
#' 
#' It would signify that tets accumulate proportionally more deleterious load  
svg("highN.modifierN.ploidy.svg")
ggplot(gen.indpop20,aes(x=MODIF_tot_minor/1e4,y=HIGH_tot_minor/1e4,fill=ploidy,color=ploidy))+
  ylab("High effect indels/1e4")+
  xlab("Weak effect indels/1e4")+
  geom_point(pch=21,size=4,alpha=1,color="black")+
  geom_smooth(method="lm")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text_repel(aes(label=pop),min.segment.length = 0, size=3,point.padding = 0.01,box.padding = 0.6,max.overlaps = 15,direction = "both",color="black")+
  th
dev.off()
#' Is this a way how to test whether the regression slopes differ?
himo0.glm<-glm(HIGH_tot_minor~MODIF_tot_minor+ploidy,family="poisson",data=gen.indpop20)
himo.glm<-glm(HIGH_tot_minor~MODIF_tot_minor*ploidy,family="poisson",data=gen.indpop20)
anova(himo0.glm,himo.glm,test="LRT")
#' The interaction of ploidy and the number of weak (MODIF)  indels is significant
#' Is it significant even if I add depth as random factor?
#' 
himo0.glmer<-glmer(HIGH_tot_minor~MODIF_tot_minor+ploidy+(1|dp),family="poisson",data=gen.indpop20)
himo.glmer<-glmer(HIGH_tot_minor~MODIF_tot_minor*ploidy+(1|dp),family="poisson",data=gen.indpop20)
anova(himo0.glmer,himo.glmer,test="LRT")
#' it fails to converge
#' if I add it as a fixed effect factor?
himo0.dp.glm<-glm(HIGH_tot_minor~MODIF_tot_minor+ploidy+dp,family="poisson",data=gen.indpop20)
himo.dp.glm<-glm(HIGH_tot_minor~MODIF_tot_minor*ploidy+dp,family="poisson",data=gen.indpop20)
anova(himo0.dp.glm,himo.dp.glm,test="LRT")
summary(himo.dp.glm)
#' this is still significant, so I will probably use this model as the final one

#' This was for minor counts, how does it look for the complete dataset?
svg("highN.modifierN.major.ploidy.svg")
ggplot(gen.indpop20,aes(x=MODIF_tot/1e4,y=HIGH_tot/1e4,fill=ploidy,color=ploidy))+
  ylab("High effect indels/1e4")+
  xlab("Weak effect indels/1e4")+
  geom_smooth(method="lm")+
  geom_point(pch=21,size=4,alpha=1,color="black")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text_repel(aes(label=pop),min.segment.length = 0, size=3,point.padding = 0.01,box.padding = 0.6,max.overlaps = 15,direction = "both",color="black")+
  th
dev.off()
#' Pretty much the same


#' What about some other fitness classes of indels
#' For these I do not have minor counts, but the pattern remains the same
svg("intronN.minor.indel.ploidy.svg")
ggplot(gen.indpop20,aes(x=ploidy,y=INTRON_tot_minor/1e4,fill=ploidy))+
  ylab("Number of intron indels/1e4")+
  geom_boxplot()+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  geom_point(pch=21,size=2,position=position_jitterdodge(),alpha=0.8)+
  geom_text(aes(label=pop),size=4,check_overlap = F,nudge_x = 0.2)+
  th
dev.off()

intron.wilcox<-wilcox.test(gen.indpop20[gen.indpop20$ploidy==2,"INTRON_tot_minor"],gen.indpop20[gen.indpop20$ploidy==4,"INTRON_tot_minor"])
print(intron.wilcox)
#' Introns show significant difference between ploidies
#' now include dp as covariate
sites_intron_dp.glm<-glm(INTRON_tot_minor~ploidy*dp,data=gen.indpop20, family="poisson")
sites_intron_dp.glm0<-glm(INTRON_tot_minor~1+dp,data=gen.indpop20, family="poisson")
anova(sites_intron_dp.glm0,sites_intron_dp.glm,test="LRT")
# it is super significant difference even if dp included

#' The correlation high vs. intron
#' 
#' This is meaningfull because introns and high effect probably share more common genomic
#' background than if I compare weak vs. high, while weak effect indels are mainly down or upstream of cds
svg("highN.intronN.ploidy.svg")
ggplot(gen.indpop20,aes(x=INTRON_tot_minor/1e4,y=HIGH_tot_minor/1e4,fill=ploidy,color=ploidy))+
  ylab("High effect indels/1e4")+
  xlab("Intron indels/1e4")+
  geom_smooth(method="lm")+
  geom_point(pch=21,size=4,alpha=1,color="black")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text_repel(aes(label=pop),min.segment.length = 0, size=3,point.padding = 0.01,box.padding = 0.6,max.overlaps = 15,direction = "both",color="black")+
  th
dev.off()

hiint.glm<-glm(HIGH_tot~INTRON_tot_minor*ploidy,family="poisson",data=gen.indpop20)
summary(hiint.glm)

hiin0.dp.glm<-glm(HIGH_tot_minor~INTRON_tot_minor+ploidy+dp,family="poisson",data=gen.indpop20)
hiin.dp.glm<-glm(HIGH_tot_minor~INTRON_tot_minor*ploidy+dp,family="poisson",data=gen.indpop20)
anova(hiin0.dp.glm,hiin.dp.glm,test="LRT")
summary(hiin.dp.glm)
#' The interaction is not significant between HIGH and INTRON, although visually the lm slopes are different
#' Maybe if I remove the depth it will become significant?
hiin0.glm<-glm(HIGH_tot_minor~INTRON_tot_minor+ploidy,family="poisson",data=gen.indpop20)
hiin.glm<-glm(HIGH_tot_minor~INTRON_tot_minor*ploidy,family="poisson",data=gen.indpop20)
anova(hiin0.glm,hiin.glm,test="LRT")
summary(hiin.glm)
#' yes this is significant 0.0199, so maybe I will use this model

#' 
#' check both putatively neutral categories: intron vs. weak effect
svg("intronN.modifierN.ploidy.svg")
ggplot(gen.indpop20,aes(x=MODIF_tot_minor/1e4,y=INTRON_tot_minor/1e4,fill=ploidy,color=ploidy))+
  ylab("Intron indels/1e4")+
  xlab("Modifier indels/1e4")+
  geom_smooth(method="lm")+
  geom_point(pch=21,size=4,alpha=1,color="black")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text_repel(aes(label=pop),min.segment.length = 0, size=3,point.padding = 0.01,box.padding = 0.6,max.overlaps = 15,direction = "both",color="black")+
  th
dev.off()
moin0.glm<-glm(INTRON_tot_minor~MODIF_tot_minor+ploidy+dp,family="poisson",data=gen.indpop20)
moin.glm<-glm(INTRON_tot_minor~MODIF_tot_minor*ploidy+dp,family="poisson",data=gen.indpop20)
anova(moin0.glm,moin.glm,test="LRT")
summary(moin.glm)


#' 
#' **The conclusion is that the effect of ploidy affects both total numbers and also accumulation**
#' **of strongly deleterious indels. Tetraploids accumulate more and more deleterious** 

##' The last chapter deals with actual load indicators as designed for human data by Simons and Sella
#' It is based on counting of genotypes
#' 

#' I'm comparing the slope similar as above, this time for high vs. weak effect heterozygous genotypes
#' Tets have higher proportion of heterozygots 1-p^2+q^2, I'm counting all dosages 0/0/0/1 0/0/1/1/ 0/1/1/1
#' I don't know if I should go into the genotype counts, it is rather confusing
#' 
#' This is the heterozygous genotypes for diploids:
#'  HIGH_het=`grep HIGH $vcf |cut -f 10-16|grep -E -o "0/1" | wc -l`
#'  And for tetraploids
#'  HIGH_het=`grep HIGH $vcf |cut -f 10-16|grep -E -o "0/0/0/1|0/1/1/1|0/0/1/1" | wc -l` 

svg("HIGH_het.MODIF_het.ploidy.svg")
ggplot(gen.indpop20,aes(x=MODIF_het/1e4,y=HIGH_het/1e4,fill=ploidy,color=ploidy))+
  ylab("High het indels/1e4")+
  xlab("Weak het indels/1e4")+
  geom_smooth(method="lm")+
  geom_point(pch=21,size=4,alpha=1,color="black")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text_repel(aes(label=pop),min.segment.length = 0, size=3,point.padding = 0.01,box.padding = 0.6,max.overlaps = 15,direction = "both",color="black")+
  th
dev.off()
#' Tetraploids have higher "concealed load" load that is in heterozygous genotypes, compared to diploids.
#' It is because they have higher heterozygosity due to HW and tetrasomic inheritance.
#' The question is whether they also accumulate proportionally more deleterious "concealed load"?
#' 
#' First lets test the difference
high.het.wilcox<-wilcox.test(gen.indpop20[gen.indpop20$ploidy==2,"HIGH_het"],gen.indpop20[gen.indpop20$ploidy==4,"HIGH_het"])
print(high.het.wilcox)

#' Now the slope without depth
het.glm0<-glm(HIGH_het~MODIF_het+ploidy,family="poisson",data=gen.indpop20)
het.glm<-glm(HIGH_het~MODIF_het*ploidy,family="poisson",data=gen.indpop20)
anova(het.glm,het.glm0,test="LRT")
summary(het.glm)
#' Yes it is significant
#' is it significant even with depth
het.dp.glm0<-glm(HIGH_het~MODIF_het+ploidy+dp,family="poisson",data=gen.indpop20)
het.dp.glm<-glm(HIGH_het~MODIF_het*ploidy+dp,family="poisson",data=gen.indpop20)
anova(het.dp.glm,het.dp.glm0,test="LRT")
summary(het.dp.glm)
#' Yes so even with depth included the interaction between ploidy and number of weak effect indels 

#' Homozygous genotypes = expressed load:
svg("HIGH_hom.MODIF_hom.ploidy.svg")
ggplot(gen.indpop20,aes(x=MODIF_hom/1e3,y=HIGH_hom/1e3,fill=ploidy,color=ploidy))+
  ylab("High hom indels/1e3")+
  xlab("Weak hom indels/1e3")+
  geom_smooth(method="lm")+
  geom_point(pch=21,size=4,alpha=1,color="black")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text_repel(aes(label=pop),min.segment.length = 0, size=3,point.padding = 0.01,box.padding = 0.6,max.overlaps = 15,direction = "both",color="black")+
  th
dev.off()
#' Here the diploids have much higher expressed load
hom.dp.glm0<-glm(HIGH_hom~MODIF_hom+ploidy+dp,family="poisson",data=gen.indpop20)
hom.dp.glm<-glm(HIGH_hom~MODIF_hom*ploidy+dp,family="poisson",data=gen.indpop20)
anova(hom.dp.glm,hom.dp.glm0,test="LRT")
summary(hom.dp.glm)
# also the interaction is significant, however diploids accumulate more, the slope is steeper for them


#' Just to show what proportion of high effect indels is in what genotype per ploidy
svg("HIGH_het.HIGH_hom.ploidy.svg")
ggplot(gen.indpop20,aes(x=HIGH_hom/1e4,y=HIGH_het/1e4,fill=ploidy))+
  ylab("High het indels/1e4")+
  xlab("High hom indels/1e4")+
  geom_point(pch=21,size=4,alpha=1)+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  
  #geom_text(aes(label=pop),size=3,check_overlap = F,nudge_x = 1)+
  #geom_smooth(method=lm)+
  th
dev.off()

svg("MODIF_het.MODIF_hom.ploidy.svg")
ggplot(gen.indpop20,aes(x=MODIF_hom/1e4,y=MODIF_het/1e4,fill=ploidy))+
  ylab("Weak het indels/1e4")+
  xlab("Weak hom indels/1e4")+
  geom_point(pch=21,size=4,alpha=1)+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  
  #geom_text(aes(label=pop),size=3,check_overlap = F,nudge_x = 1)+
  #geom_smooth(method=lm)+
  th
dev.off()

#' Plot additive load, for this index I summed upped all alleles present in population
#' HIGH_additive=`echo "$HIGH_het+$HIGH_hom*2" |bc` this is for diploid
#' And this for tetraploid
#' HIGH_het1=`grep HIGH $vcf |cut -f 10-16|grep -E -o "0/0/0/1" | wc -l`  dosage1
#' HIGH_het2=`grep HIGH $vcf |cut -f 10-16|grep -E -o "0/0/1/1" | wc -l`  dosage2
#' HIGH_het3=`grep HIGH $vcf |cut -f 10-16|grep -E -o "0/1/1/1" | wc -l`  dosage3
#' HIGH_additive=`echo "$HIGH_het1+$HIGH_het2*2+$HIGH_het3*3+$HIGH_hom*4" |bc`


svg("HIGH_additive.WEAK_additive.ploidy.svg")
ggplot(gen.indpop20,aes(y=HIGH_additive/1e4,x=MODIF_additive/1e4,fill=ploidy,color=ploidy))+
  ylab("Strong effect indel alleles/1e4")+
  xlab("Weak effect indel alleles/1e4")+
  geom_smooth(method="lm")+
  geom_point(pch=21,size=4,alpha=1,color="black")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text_repel(aes(label=pop),min.segment.length = 0, size=3,point.padding = 0.01,box.padding = 0.6,max.overlaps = 15,direction = "both",color="black")+
  th
dev.off()
#' If the indels are additive the genetic load is higher in tetraploids
#' If the indels are recessive the genetic load is higher in diploids
#' For the additive load it seems it follows the same line as total numbers 
#' In the sense that tetraploids accumulate proportionally more deleterious load



#' one idea maybe just for visualization divide the number of allele by 2
#' to see the difference in the slope
gen.indpop20$ploidy
str(as.numeric(gen.indpop20$HIGH_additive/2))
str(gen.indpop20$HIGH_additive)
#there is probably some dplyr solution
gen.indpop20<-gen.indpop20 %>% 
  mutate(HIGH_additive_scaled = ifelse(ploidy == '4', HIGH_additive / 2, HIGH_additive),
         MODIF_additive_scaled = ifelse(ploidy == '4', MODIF_additive / 2, MODIF_additive))

  #mutate(HIGH_additive_scaled = case_when(ploidy == '4' ~ as.numeric(HIGH_additive/2),
   #                                       .default = as.numeric(HIGH_additive))) # this was giving an error
                                          
svg("HIGH_additive.MODIF_additive.scaled.tets.ploidy.svg")
ggplot(gen.indpop20,aes(y=HIGH_additive_scaled/1e4,x=MODIF_additive_scaled/1e4,fill=ploidy,color=ploidy))+
  ylab("Weak indel alleles/1e4")+
  xlab("High indel alleles/1e4")+
  geom_smooth(method="lm")+
  geom_point(pch=21,size=4,alpha=1,color="black")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text_repel(aes(label=pop),min.segment.length = 0, size=3,point.padding = 0.01,box.padding = 0.6,max.overlaps = 15,direction = "both",color="black")+
  th
dev.off()
#' OK the difference is clear I think this is also one of the main picture that should go to the text
#' Let's test the significance
additive.dp.glm0<-glm(HIGH_additive~MODIF_additive+ploidy+dp,family="poisson",data=gen.indpop20)
additive.dp.glm<-glm(HIGH_additive~MODIF_additive*ploidy+dp,family="poisson",data=gen.indpop20)
anova(additive.dp.glm,additive.dp.glm0,test="LRT")
summary(additive.dp.glm)
# This is significant even with depth as a co-variate.


#' final figure for publcation working on this 23.10.23 on mypc
library(ggrepel)

th<-theme(axis.title = element_text(size=23),
          axis.text.y=element_text(size=23,color="black"),
          axis.text.x=element_text(size=23,color="black"),
          plot.title=element_text(size=25),
          legend.text = element_text(size=18),
          strip.text.x = element_text(size = 18),
          panel.background = element_rect(fill = "white", colour = "white"),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(color="black",linetype="solid")
)


gen.indpop20$ploidy
gen.indpop20$scaled_MODIF_tot_minor <-scale(gen.indpop20$MODIF_tot_minor)
gen.indpop20$scaled_HIGH_tot_minor <-scale(gen.indpop20$HIGH_tot_minor)

slodip<-coef(lm(scaled_HIGH_tot_minor~scaled_MODIF_tot_minor,data=gen.indpop20[gen.indpop20$ploidy == 2,]))[2]
slotet<-coef(lm(scaled_HIGH_tot_minor~scaled_MODIF_tot_minor,data=gen.indpop20[gen.indpop20$ploidy == 4,]))[2]
slodip<-round(slodip,digits=3)
slotet<-round(slotet,digits=3) # why is the slope smaller in tetraploids? even after scaling the data


lm1<-lm(MODIF_tot_minor~HIGH_tot_minor*ploidy, data=gen.indpop20)
coef(lm1)
# in the introns the result is as we want
lm2<-lm(MODIF_tot_minor~INTRON_tot_minor*ploidy, data=gen.indpop20)
coef(lm2) # also negative interaction

summary(lm2)


gen.indpop20$scaled_INTRON_tot_minor <-scale(gen.indpop20$INTRON_tot_minor)
gen.indpop20$scaled_HIGH_tot_minor <-scale(gen.indpop20$HIGH_tot_minor)

coef(lm(scaled_INTRON_tot_minor~scaled_HIGH_tot_minor,data=gen.indpop20[gen.indpop20$ploidy == 2,]))[2]
coef(lm(scaled_INTRON_tot_minor~scaled_HIGH_tot_minor,data=gen.indpop20[gen.indpop20$ploidy == 4,]))[2]

library(scales)
fig2c<-ggplot(gen.indpop20,aes(x=MODIF_tot_minor,y=HIGH_tot_minor,fill=ploidy,color=ploidy))+
  ylab("Strong effect indels")+
  xlab("Weak effect indels")+ 
  geom_smooth(method="glm",linewidth=1,method.args = list(family = "gaussian"))+
  geom_point(pch=21,size=4,alpha=0.7,color="black")+
  scale_x_continuous(breaks = breaks_pretty(n=4),label = label_number(scale = 1e-6,suffix = "M"))+
  scale_y_continuous(breaks = breaks_pretty(n=4),label = label_number(scale = 1e-3,suffix = "K"))+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  geom_text_repel(aes(label=pop),min.segment.length = 0, size=3,point.padding = 0.01,box.padding = 0.6,max.overlaps = 15,direction = "both",color="black")+
  #geom_text(aes(1.9e6,50e3,label=paste("\u03B2","=",as.character(slodip),sep="")),color="#1e90ff",size=7)+
  #geom_text(aes(2.1e6,50e3,label=paste("/",as.character(slotet),sep="")),color="#ffa500",size=7)+
  th


svg("fig2c.svg")
print(fig2c)
dev.off()

saveRDS(fig2c,"indel.load.fig.rds") # eventually you will put all the figures together






fig2b<-ggplot(gen.indpop20,aes(y=HIGH_additive_scaled/1e4,x=MODIF_additive_scaled/1e4,fill=ploidy,colour=ploidy))+
  ylab("Strong alleles/1e4")+
  xlab("Weak alleles/1e4")+
  #ggtitle("Alleles")+
  geom_smooth(method=lm,size=0.5)+
  geom_point(pch=21,size=2,alpha=0.5,color="black")+
  scale_fill_manual(values=c("#1e90ff","#ffa500"))+
  scale_color_manual(values=c("#1e90ff","#ffa500"))+
  theme(axis.text.x=element_text(size = 10),
        axis.title = element_text(size=10),
        axis.text.y=element_text(size=10),
        legend.text = element_blank(),
        legend.position = "none")

svg("main.figure.indel.load.svg")
fig2a+ annotation_custom(ggplotGrob(fig2b), xmin = 180, xmax = 260, 
                         ymin = 3.5, ymax = 6)

dev.off()

indel.load.fig<-fig2a+ annotation_custom(ggplotGrob(fig2b), xmin = 180, xmax = 260, 
                                         ymin = 3.5, ymax = 6)


png("main.figure.indel.load.png")
indel.load.fig
dev.off()



##' Now let's generate site frequency spectrum of diploid and tetraploid
#' It will be spectrum from minor counts only and separate for diploid and tetraploid
#' The SFS is biased even in some of the pop20 populations. In some scaffolds 
#' doubletons are the most frequent category especially in low covered tetraploid populations 
#' such as ZIT, PAT, PER. Well let's try to sum it up across genome, because now I have it per scaffold
#' 
#' /ERC/arenosa_mapped_to_arenosa/filtering_results/indels/pop.8x.ac0.indels.annotated/vcfs.from.hom.het.counts/minor.concat/sfs$ parallel --bar -j 10 "mv  {}*.sfs pop20" :::: pop20.txt 
#' tar -cvf pop20.sfs.tar pop20/
#' scp filip_kolar@storage-brno12-cerit.metacentrum.cz:brno3/ERC/arenosa_mapped_to_arenosa/filtering_results/indels/pop.8x.ac0.indels.annotated/vcfs.from.hom.het.counts/minor.concat/sfs/pop20.sfs.tar . 
#' parallel "cat {}*high.sfs > {}.high.sfs" :::: ../pop20.txt this generates genome wide sfs from scaffolds
#' parallel "cat {}*modifier.sfs > {}.modifier.sfs" :::: ../pop20.tx
#' 

pop20<-as.character(gen.indpop20$pop)
pop<-as.character(gen.indpop$pop)

summary(gen.indpop20$nind)
write.table(pop20,"pop20.txt",quote=F,sep="\t",col.names=F,row.names = F)
pop="BAB"
# for the type you must specify folder with sfs, either minor/subset etc.

plotsfs<- function (pop,type) {
  # HIGH effect sites sfs
  sfsHIGH<- read.table(paste(type,"/",pop,".high.sfs",sep=""),col.names = c("No.Indels","Frequency"))
  sfsHIGH<-sfsHIGH %>% group_by(Frequency) %>% summarise (No.Indels = sum(No.Indels)) # sum the frequencies across genome
  sfsHIGH$prop.Indels<-sfsHIGH$No.Indels/sum(sfsHIGH$No.Indels)
  sfsHIGH$cat<-rep("High",dim(sfsHIGH)[1])
  # WEAK effect sites sfs
  sfsMODIFIER<- read.table(paste(type,"/",pop,".modifier.sfs",sep=""),col.names = c("No.Indels","Frequency"))
  sfsMODIFIER<-sfsMODIFIER %>% group_by(Frequency) %>% summarise (No.Indels = sum(No.Indels)) # sum the frequencies across genome
  sfsMODIFIER$prop.Indels<-sfsMODIFIER$No.Indels/sum(sfsMODIFIER$No.Indels)
  sfsMODIFIER$cat<-rep("Weak",dim(sfsMODIFIER)[1])
  
  # INTRON weak effect sites sfs
  sfsintron<- read.table(paste(type,"/",pop,".intron.sfs",sep=""),col.names = c("No.Indels","Frequency"))
  sfsintron<-sfsintron %>% group_by(Frequency) %>% summarise (No.Indels = sum(No.Indels)) # sum the frequencies across genome
  sfsintron$prop.Indels<-sfsintron$No.Indels/sum(sfsintron$No.Indels)
  sfsintron$cat<-rep("Intron",dim(sfsintron)[1])
  
  sfs.eff<-rbind(sfsHIGH,sfsMODIFIER,sfsintron)
  sfs.eff$Frequency<-as.factor(sfs.eff$Frequency)
  sfs.eff$pop<-rep(pop,dim(sfs.eff)[1])
  th<-theme(axis.text.x=element_text(size=15),
            axis.title = element_text(size=23),
            axis.text.y=element_text(size=23),plot.title=element_text(size=25),
            legend.text = element_text(size=18))
  ggplot(sfs.eff,aes(x=Frequency,y=prop.Indels,fill=cat))+
    geom_bar(position = "dodge",stat="identity")+
    #geom_text(aes(x=4,y=0.2,label=as.character(sum(sfsHIGH$No.Indels))),colour="#F8766D")+ 
    #geom_text(aes(x=4,y=0.15,label=as.character(sum(sfsMODIFIER$No.Indels))),colour="#619CFF")+
    ggtitle(pop)+
    th
  return(sfs.eff) # switch this on if you want to generate the diploid tetraploid sfs
  
}
plotsfs("BAB","minor.sfs")

#pdf("sfs.subset.allpop.High.Intron.Weak.pdf",onefile = T)
#lapply(pop,plotsfs)
#dev.off()
#'**full sfs from the pop20**
pdf("sfs.subset.pop20.High.Intron.Weak.pdf",onefile = T)
lapply(pop20,plotsfs,type="subset.sfs")
dev.off()

#' the sfs are perfect always the high category has excess of singletons and deficiency of 
#' higher frequencies

#' for the final publication I will use minor spectra where only alleles with freq 0.5 and lower were used
#' 
#' scp filip_kolar@storage-brno12-cerit.metacentrum.cz:brno3/ERC/arenosa_mapped_to_arenosa/filtering_results/indels/pop.5x.ac0.indels.annotated/vcfs.from.hom.het.counts/minor.sfs/minor/pop.sfs/* .

pdf("sfs.minor.pop20.High.Intron.Weak.pdf",onefile = T)
lapply(pop20,plotsfs,type="minor.sfs")
dev.off()


# now I need to generate average tetraploid sfs and average diploid sfs
# to do that I need matrix of all tet and dip sfs
plotsfs("BAB")
pop20.dip<-as.character(gen.indpop20[gen.indpop20$ploidy == 2,"pop"])
pop20.tet<-as.character(gen.indpop20[gen.indpop20$ploidy == 4,"pop"])

# before this you must switch on the return function in the plotsfs function
dip.sfs<-bind_rows(lapply(pop20.dip,plotsfs,type="minor.sfs"))
tet.sfs<-bind_rows(lapply(pop20.tet,plotsfs,type="minor.sfs"))

dip.sfs$Frequency<-as.factor((dip.sfs$Frequency))
tet.sfs$Frequency<-as.factor((tet.sfs$Frequency))

svg("diploid.minor.points.sfs.svg",width=10)
ggplot(dip.sfs,aes(x=Frequency,y=prop.Indels,fill=cat))+
  geom_boxplot()+
  #geom_point(pch=21,size=2,alpha=0.5,position = position_jitterdodge())+
  ggtitle("diploid")+
  scale_y_continuous(limits = c(0,0.45))+
  th
dev.off()

svg("tetraploid.minor.points.sfs.svg",width=10)
ggplot(tet.sfs,aes(x=Frequency,y=prop.Indels,fill=cat))+
  ggtitle("tetraploid")+
  scale_y_continuous(limits = c(0,0.45))+
  geom_boxplot()+
  #geom_point(pch=21,size=2,alpha=0.5,position = position_jitterdodge())+
  th
dev.off()

#' try to plot diploids and tetraploids in a common sfs plot
#' Use just high and weak, there will be four categories, High2x, High4x, Weak2x, High4x
head(dip.sfs)
head(tet.sfs)
dip.sfs$ep<-paste(dip.sfs$cat,"2x",sep=".")
tet.sfs$ep<-paste(tet.sfs$cat,"4x",sep=".")
all.sfs<-rbind(dip.sfs,tet.sfs)

svg("all.minor.points.sfs.svg",width=10)
all.sfs%>%
  filter(cat != "Intron") %>%
  mutate(ep = fct_relevel(ep,"High.2x","Weak.2x","High.4x","Weak.4x")) %>%
  ggplot(aes(x=Frequency,y=prop.Indels,fill=ep))+
  scale_fill_manual(values=c("#F8766D","orange","#619CFF","#00BFC4"))+
  #ggtitle("tetraploid")+
  scale_y_continuous(limits = c(0,0.45))+
  scale_x_discrete(limits = c(1:6))+
  geom_boxplot()+
  geom_point(pch=21,size=2,alpha=0.5,position = position_jitterdodge())+
  th
dev.off()

#' excess of singleton in High effect class
